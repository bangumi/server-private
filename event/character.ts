import { getSlimCacheKey } from '@app/lib/character/cache';
import redis from '@app/lib/redis.ts';

import { EventOp } from './type';

interface CharacterKey {
  crt_id: number;
}

interface Payload {
  op: EventOp;
}

export async function handle(key: string, value: string) {
  const idx = JSON.parse(key) as CharacterKey;
  const payload = JSON.parse(value) as Payload;
  switch (payload.op) {
    case EventOp.Create: {
      break;
    }
    case EventOp.Update:
    case EventOp.Delete: {
      await redis.del(getSlimCacheKey(idx.crt_id));
      break;
    }
    case EventOp.Snapshot: {
      break;
    }
  }
}
