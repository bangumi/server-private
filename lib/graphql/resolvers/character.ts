import type { Wiki } from '@bgm38/wiki';
import { parse as parseWiki, WikiSyntaxError } from '@bgm38/wiki';

import type * as entity from '@app/lib/orm/entity/index.ts';
import { personImages } from '@app/lib/response.ts';

export function convertCharacter(character: entity.Character) {
  let wiki: Wiki = {
    type: '',
    data: [],
  };
  try {
    wiki = parseWiki(character.infobox);
  } catch (error) {
    if (!(error instanceof WikiSyntaxError)) {
      throw error;
    }
  }
  const infobox = wiki.data.map((item) => {
    if (item.array) {
      return item;
    }
    return {
      key: item.key,
      values: [
        {
          v: item.value,
        },
      ],
    };
  });
  return {
    id: character.id,
    name: character.name,
    role: character.role,
    infobox,
    summary: character.summary,
    images: personImages(character.img),
    comment: character.comment,
    collects: character.collects,
    last_post: character.lastPost,
    lock: character.lock,
    redirect: character.redirect,
    nsfw: character.nsfw,
  };
}
