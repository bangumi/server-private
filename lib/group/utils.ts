import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
