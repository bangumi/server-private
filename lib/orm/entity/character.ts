import { Column, Entity, Index, PrimaryColumn, PrimaryGeneratedColumn } from 'typeorm';

import type { Person } from './person.ts';
import type { Subject } from './subject.ts';

@Index('crt_role', ['role'], {})
@Index('crt_lock', ['lock'], {})
@Index('crt_ban', ['ban'], {})
@Entity('chii_characters', { schema: 'bangumi' })
export class Character {
  @PrimaryGeneratedColumn({ type: 'mediumint', name: 'crt_id', unsigned: true })
  id!: number;

  @Column('varchar', { name: 'crt_name', length: 255 })
  name!: string;

  @Column('tinyint', {
    name: 'crt_role',
    comment: '角色，机体，组织。。',
    unsigned: true,
  })
  role!: number;

  @Column('mediumtext', { name: 'crt_infobox' })
  infobox!: string;

  @Column('mediumtext', { name: 'crt_summary' })
  summary!: string;

  @Column('varchar', { name: 'crt_img', length: 255 })
  img!: string;

  @Column('mediumint', {
    name: 'crt_comment',
    unsigned: true,
    default: () => "'0'",
  })
  comment!: number;

  @Column('mediumint', { name: 'crt_collects', unsigned: true })
  collects!: number;

  @Column('int', { name: 'crt_dateline', unsigned: true })
  updatedAt!: number;

  @Column('int', { name: 'crt_lastpost', unsigned: true })
  lastPost!: number;

  @Column('tinyint', { name: 'crt_lock', default: () => "'0'" })
  lock!: number;

  @Column('varchar', { name: 'crt_img_anidb', length: 255 })
  imgAnidb!: string;

  @Column('mediumint', { name: 'crt_anidb_id', unsigned: true })
  anidbID!: number;

  @Column('tinyint', { name: 'crt_ban', unsigned: true, default: () => "'0'" })
  ban!: number;

  @Column('int', { name: 'crt_redirect', unsigned: true, default: () => "'0'" })
  redirect!: number;

  @Column('tinyint', { name: 'crt_nsfw', unsigned: true, width: 1 })
  nsfw!: boolean;
}

@Index('subject_id', ['subjectID'], {})
@Index('crt_type', ['type'], {})
@Index('subject_type_id', ['subjectTypeID'], {})
@Entity('chii_crt_subject_index', { schema: 'bangumi' })
export class CharacterSubjects {
  @PrimaryColumn('mediumint', { primary: true, name: 'crt_id', unsigned: true })
  characterID!: number;

  @PrimaryColumn('mediumint', { primary: true, name: 'subject_id', unsigned: true })
  subjectID!: number;

  @Column('tinyint', { name: 'subject_type_id', unsigned: true })
  subjectTypeID!: number;

  @Column('tinyint', {
    name: 'crt_type',
    comment: '主角，配角',
    unsigned: true,
  })
  type!: number;

  // @Column('mediumtext', {
  //   name: 'ctr_appear_eps',
  //   comment: '可选，角色出场的的章节',
  // })
  // appearEps: string;

  @Column('tinyint', { name: 'crt_order', unsigned: true })
  order!: number;

  character!: Character;

  subject!: Subject;
}

@Index('prsn_id', ['personID'], {})
@Index('subject_id', ['subjectID'], {})
@Index('subject_type_id', ['subjectTypeID'], {})
@Entity('chii_crt_cast_index', { schema: 'bangumi' })
export class Cast {
  @PrimaryColumn('mediumint', { primary: true, name: 'crt_id', unsigned: true })
  characterID!: number;

  @PrimaryColumn('mediumint', { primary: true, name: 'prsn_id', unsigned: true })
  personID!: number;

  @PrimaryColumn('mediumint', { primary: true, name: 'subject_id', unsigned: true })
  subjectID!: number;

  @Column('tinyint', {
    name: 'subject_type_id',
    comment: '根据人物归类查询角色，动画，书籍，游戏',
    unsigned: true,
  })
  subjectTypeID!: number;

  @Column('varchar', {
    name: 'summary',
    comment: '幼年，男乱马，女乱马，变身形态，少女形态。。',
    length: 255,
  })
  summary!: string;

  character!: Character;
  person!: Person;
  subject!: Subject;
}
