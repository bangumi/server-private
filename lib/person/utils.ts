import { db, op, schema } from '@app/drizzle';

import type { PersonCat } from './type';

export async function getPersonCollect(
  cat: PersonCat,
  userID: number,
  personID: number,
): Promise<number | null> {
  const [d] = await db
    .select()
    .from(schema.chiiPersonCollects)
    .where(
      op.and(
        op.eq(schema.chiiPersonCollects.cat, cat),
        op.eq(schema.chiiPersonCollects.uid, userID),
        op.eq(schema.chiiPersonCollects.mid, personID),
      ),
    )
    .limit(1);
  if (!d) {
    return null;
  }
  return d.createdAt;
}
