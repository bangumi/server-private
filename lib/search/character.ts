import { client, type SearchResult } from '@app/lib/search/client.ts';
import type * as req from '@app/lib/types/req.ts';

const index = client.index('characters');

interface SearchRequest {
  keyword: string;
  filter?: req.ICharacterSearchFilter;
  limit: number;
  offset: number;
}

function convertFilter(filter?: req.ICharacterSearchFilter): string[][] {
  const filters: string[][] = [];
  if (!filter) {
    return filters;
  }
  if (filter.nsfw !== undefined) {
    filters.push([`nsfw = ${filter.nsfw}`]);
  }
  return filters;
}

export async function search(request: SearchRequest): Promise<SearchResult> {
  const filter = convertFilter(request.filter);
  const results = await index.search(request.keyword, {
    filter,
    limit: request.limit,
    offset: request.offset,
  });
  return {
    ids: results.hits.map((hit) => hit.id as number),
    total: results.estimatedTotalHits,
  };
}
