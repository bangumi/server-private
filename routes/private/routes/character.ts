import { Type as t } from '@sinclair/typebox';

import { db, op } from '@app/drizzle/db.ts';
import type * as orm from '@app/drizzle/orm.ts';
import * as schema from '@app/drizzle/schema';
import { NotFoundError } from '@app/lib/error.ts';
import { Security, Tag } from '@app/lib/openapi/index.ts';
import * as convert from '@app/lib/types/convert.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import * as req from '@app/lib/types/req.ts';
import * as res from '@app/lib/types/res.ts';
import { formatErrors } from '@app/lib/types/res.ts';
import type { App } from '@app/routes/type.ts';

function toCharacterSubject(
  subject: orm.ISubject,
  relation: orm.ICharacterSubject,
  actors: res.ISlimPerson[],
): res.ICharacterSubject {
  return {
    subject: convert.toSlimSubject(subject),
    actors: actors,
    type: relation.type,
  };
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.get(
    '/characters/:characterID',
    {
      schema: {
        summary: '获取角色',
        operationId: 'getCharacter',
        tags: [Tag.Character],
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        params: t.Object({
          characterID: t.Integer(),
        }),
        response: {
          200: t.Ref(res.Character),
          404: t.Ref(res.Error, {
            'x-examples': formatErrors(new NotFoundError('character')),
          }),
        },
      },
    },
    async ({ auth, params: { characterID } }) => {
      const data = await db
        .select()
        .from(schema.chiiCharacters)
        .where(
          op.and(
            op.eq(schema.chiiCharacters.id, characterID),
            op.ne(schema.chiiCharacters.ban, 1),
            auth.allowNsfw ? undefined : op.eq(schema.chiiCharacters.nsfw, false),
          ),
        )
        .execute();
      for (const d of data) {
        return convert.toCharacter(d);
      }
      throw new NotFoundError(`character ${characterID}`);
    },
  );

  app.get(
    '/characters/:characterID/casts',
    {
      schema: {
        summary: '获取角色出演作品',
        operationId: 'getCharacterCasts',
        tags: [Tag.Character],
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        params: t.Object({
          characterID: t.Integer(),
        }),
        querystring: t.Object({
          subjectType: t.Optional(t.Ref(req.SubjectType)),
          type: t.Optional(t.Integer({ description: '角色出场类型: 主角，配角，客串' })),
          limit: t.Optional(
            t.Integer({ default: 20, minimum: 1, maximum: 100, description: 'max 100' }),
          ),
          offset: t.Optional(t.Integer({ default: 0, minimum: 0, description: 'min 0' })),
        }),
        response: {
          200: res.Paged(t.Ref(res.CharacterSubject)),
          404: t.Ref(res.Error, {
            'x-examples': formatErrors(new NotFoundError('character')),
          }),
        },
      },
    },
    async ({
      auth,
      params: { characterID },
      query: { subjectType, type, limit = 20, offset = 0 },
    }) => {
      const character = await fetcher.fetchSlimCharacterByID(characterID, auth.allowNsfw);
      if (!character) {
        throw new NotFoundError(`character ${characterID}`);
      }
      const condition = op.and(
        op.eq(schema.chiiCharacterSubjects.characterID, characterID),
        subjectType ? op.eq(schema.chiiCharacterSubjects.subjectType, subjectType) : undefined,
        type ? op.eq(schema.chiiCharacterSubjects.type, type) : undefined,
        op.ne(schema.chiiSubjects.ban, 1),
        auth.allowNsfw ? undefined : op.eq(schema.chiiSubjects.nsfw, false),
      );
      const [{ count = 0 } = {}] = await db
        .select({ count: op.count() })
        .from(schema.chiiCharacterSubjects)
        .innerJoin(
          schema.chiiSubjects,
          op.eq(schema.chiiCharacterSubjects.subjectID, schema.chiiSubjects.id),
        )
        .where(condition)
        .execute();
      const data = await db
        .select()
        .from(schema.chiiCharacterSubjects)
        .innerJoin(
          schema.chiiSubjects,
          op.eq(schema.chiiCharacterSubjects.subjectID, schema.chiiSubjects.id),
        )
        .where(condition)
        .orderBy(
          op.asc(schema.chiiCharacterSubjects.type),
          op.asc(schema.chiiCharacterSubjects.order),
        )
        .limit(limit)
        .offset(offset)
        .execute();
      const subjectIDs = data.map((d) => d.chii_subjects.id);
      const casts = await fetcher.fetchCastsByCharacterAndSubjectIDs(
        characterID,
        subjectIDs,
        auth.allowNsfw,
      );
      const subjects = data.map((d) =>
        toCharacterSubject(
          d.chii_subjects,
          d.chii_crt_subject_index,
          casts[d.chii_subjects.id] || [],
        ),
      );
      return {
        data: subjects,
        total: count,
      };
    },
  );

  app.get(
    '/characters/:characterID/collects',
    {
      schema: {
        summary: '获取角色的收藏用户',
        operationId: 'getCharacterCollects',
        tags: [Tag.Character],
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        params: t.Object({
          characterID: t.Integer(),
        }),
        querystring: t.Object({
          limit: t.Optional(
            t.Integer({ default: 20, minimum: 1, maximum: 100, description: 'max 100' }),
          ),
          offset: t.Optional(t.Integer({ default: 0, minimum: 0, description: 'min 0' })),
        }),
        response: {
          200: res.Paged(t.Ref(res.PersonCollect)),
          404: t.Ref(res.Error, {
            'x-examples': formatErrors(new NotFoundError('character')),
          }),
        },
      },
    },
    async ({ auth, params: { characterID }, query: { limit = 20, offset = 0 } }) => {
      const character = await fetcher.fetchSlimCharacterByID(characterID, auth.allowNsfw);
      if (!character) {
        throw new NotFoundError(`character ${characterID}`);
      }
      const condition = op.and(
        op.eq(schema.chiiPersonCollects.cat, 'crt'),
        op.eq(schema.chiiPersonCollects.mid, characterID),
      );
      const [{ count = 0 } = {}] = await db
        .select({ count: op.count() })
        .from(schema.chiiPersonCollects)
        .innerJoin(schema.chiiUsers, op.eq(schema.chiiPersonCollects.uid, schema.chiiUsers.id))
        .where(condition)
        .execute();
      const data = await db
        .select()
        .from(schema.chiiPersonCollects)
        .innerJoin(schema.chiiUsers, op.eq(schema.chiiPersonCollects.uid, schema.chiiUsers.id))
        .where(condition)
        .orderBy(op.desc(schema.chiiPersonCollects.createdAt))
        .limit(limit)
        .offset(offset)
        .execute();
      const users = data.map((d) =>
        convert.toPersonCollect(d.chii_members, d.chii_person_collects),
      );
      return {
        total: count,
        data: users,
      };
    },
  );
}
