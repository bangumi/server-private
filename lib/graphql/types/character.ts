import type { Wiki } from '@bgm38/wiki';
import { parse as parseWiki, WikiSyntaxError } from '@bgm38/wiki';
import { extendType, intArg, nonNull, objectType } from 'nexus';

import type { Context } from '@app/lib/graphql/context.ts';
import * as entity from '@app/lib/orm/entity/index.ts';
import { CastRepo, CharacterRepo, CharacterSubjectsRepo } from '@app/lib/orm/index.ts';
import { personImages } from '@app/lib/response.ts';

import { Images, InfoboxItem } from './common.ts';
import { convertPerson } from './person.ts';
import { convertSubject } from './subject.ts';

const CharacterRelatedSubject = objectType({
  name: 'CharacterRelatedSubject',
  definition(t) {
    t.nonNull.field('subject', {
      type: 'Subject',
    });
    t.nonNull.int('type');
    t.nonNull.int('order');
  },
});

const CharacterRelatedPerson = objectType({
  name: 'CharacterRelatedPerson',
  definition(t) {
    t.nonNull.field('person', {
      type: 'Person',
    });
    t.nonNull.field('subject', {
      type: 'Subject',
    });
    t.nonNull.string('summary');
  },
});

const Character = objectType({
  name: 'Character',
  definition(t) {
    t.nonNull.int('id');
    t.nonNull.string('name');
    t.nonNull.int('role');
    t.list.nonNull.field('infobox', { type: InfoboxItem });
    t.nonNull.string('summary');
    t.nullable.field('images', { type: Images });
    t.nonNull.int('comment');
    t.nonNull.int('collects');
    t.nonNull.int('last_post');
    t.nonNull.int('lock');
    t.nonNull.int('redirect');
    t.nonNull.boolean('nsfw');
    t.list.nonNull.field('subjects', {
      type: CharacterRelatedSubject,
      args: {
        limit: nonNull(intArg({ default: 10 })),
        offset: nonNull(intArg({ default: 0 })),
      },
      async resolve(
        parent: { id: number },
        { limit, offset }: { limit: number; offset: number },
        { auth: { allowNsfw } }: Context,
      ) {
        let query = CharacterSubjectsRepo.createQueryBuilder('r')
          .innerJoinAndMapOne('r.subject', entity.Subject, 's', 's.id = r.subjectID')
          .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
          .where('r.characterID = :id', { id: parent.id });
        if (!allowNsfw) {
          query = query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
        }
        const relations = await query
          .orderBy('r.type', 'ASC')
          .orderBy('r.order', 'ASC')
          .skip(offset)
          .take(limit)
          .getMany();
        return relations.map((r) => ({
          subject: convertSubject(r.subject),
          type: r.type,
          order: r.order,
        }));
      },
    });
    t.list.nonNull.field('persons', {
      type: CharacterRelatedPerson,
      args: {
        limit: nonNull(intArg({ default: 10 })),
        offset: nonNull(intArg({ default: 0 })),
      },
      async resolve(
        parent: { id: number },
        { limit, offset }: { limit: number; offset: number },
        { auth: { allowNsfw } }: Context,
      ) {
        const query = CastRepo.createQueryBuilder('r')
          .innerJoinAndMapOne('r.person', entity.Person, 'p', 'p.id = r.personID')
          .innerJoinAndMapOne('r.subject', entity.Subject, 's', 's.id = r.subjectID')
          .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
          .where('r.characterID = :id', { id: parent.id });
        if (!allowNsfw) {
          query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
        }
        const relations = await query.skip(offset).take(limit).getMany();
        return relations.map((r) => ({
          person: convertPerson(r.person),
          subject: convertSubject(r.subject),
          summary: r.summary,
        }));
      },
    });
  },
});

export function convertCharacter(character: entity.Character) {
  let wiki: Wiki = {
    type: '',
    data: [],
  };
  try {
    wiki = parseWiki(character.infobox);
  } catch (error) {
    if (!(error instanceof WikiSyntaxError)) {
      throw error;
    }
  }
  const infobox = wiki.data.map((item) => {
    if (item.array) {
      return item;
    }
    return {
      key: item.key,
      values: [
        {
          v: item.value,
        },
      ],
    };
  });
  return {
    id: character.id,
    name: character.name,
    role: character.role,
    infobox,
    summary: character.summary,
    images: personImages(character.img),
    comment: character.comment,
    collects: character.collects,
    last_post: character.lastPost,
    lock: character.lock,
    redirect: character.redirect,
    nsfw: character.nsfw,
  };
}

const CharacterByIDQuery = extendType({
  type: 'Query',
  definition(t) {
    t.nonNull.field('character', {
      type: 'Character',
      args: { id: nonNull(intArg()) },
      async resolve(_parent, { id }: { id: number }, { auth: { allowNsfw } }: Context) {
        let query = CharacterRepo.createQueryBuilder('c').where('c.id = :id', { id });
        if (!allowNsfw) {
          query = query.andWhere('c.nsfw = :allowNsfw', { allowNsfw });
        }
        query = query.andWhere('c.ban = 0');
        const character = await query.getOne();
        if (!character) {
          return null;
        }
        return convertCharacter(character);
      },
    });
  },
});

export default [Character, CharacterByIDQuery];
