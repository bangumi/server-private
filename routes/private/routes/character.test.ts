import { describe, expect, test } from 'vitest';

import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './character.ts';
import { emptyAuth } from '@app/lib/auth/index.ts';

describe('character', () => {
  test('should get character', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/characters/32',
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get character with collectedAt', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 6162,
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/characters/32',
    });
    expect(res.statusCode).toBe(200);
    const character = res.json();
    expect(character.collectedAt).toEqual(1296496277);
  });

  test('should get character casts', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/characters/32/casts',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get character collects', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/characters/32/collects',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });
});
