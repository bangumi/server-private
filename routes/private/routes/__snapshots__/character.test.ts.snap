// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`character > should get character 1`] = `
Object {
  "collects": 329,
  "comment": 56,
  "id": 32,
  "images": Object {
    "grid": "https://lain.bgm.tv/pic/crt/g/ce/65/32_crt_0g9f9.jpg?r=1606033650",
    "large": "https://lain.bgm.tv/pic/crt/l/ce/65/32_crt_0g9f9.jpg?r=1606033650",
    "medium": "https://lain.bgm.tv/pic/crt/m/ce/65/32_crt_0g9f9.jpg?r=1606033650",
    "small": "https://lain.bgm.tv/pic/crt/s/ce/65/32_crt_0g9f9.jpg?r=1606033650",
  },
  "info": "性别 女 / 生日 4月13日 / 身高 152cm",
  "infobox": Array [
    Object {
      "key": "生日",
      "values": Array [
        Object {
          "v": "4月13日",
        },
      ],
    },
    Object {
      "key": "身高",
      "values": Array [
        Object {
          "v": "152cm",
        },
      ],
    },
    Object {
      "key": "性别",
      "values": Array [
        Object {
          "v": "女",
        },
      ],
    },
    Object {
      "key": "别名",
      "values": Array [
        Object {
          "k": "日文名",
          "v": "ちぃ",
        },
        Object {
          "k": "罗马字",
          "v": "chii",
        },
        Object {
          "k": "第二中文名",
          "v": "小唧",
        },
      ],
    },
    Object {
      "key": "引用来源",
      "values": Array [
        Object {
          "v": "anidb.net",
        },
      ],
    },
    Object {
      "key": "简体中文名",
      "values": Array [
        Object {
          "v": "小叽",
        },
      ],
    },
    Object {
      "key": "忌日",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "血型",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "体重",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "BWH",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "声优",
      "values": Array [
        Object {
          "v": "田中理惠",
        },
      ],
    },
  ],
  "lock": false,
  "name": "ちぃ",
  "nameCN": "小叽",
  "nsfw": false,
  "redirect": 0,
  "role": 1,
  "summary": "传说中性能超强的人型电脑，故事第一话时被人弃置在垃圾场，后被我们的本须和秀树发现，并抱回家。由于开始时唧只会'唧，唧'的这样叫，所以秀树为其取名 '唧'TV版第二话「ちぃでかける」时发现小唧本身并没有安OS，不过因为拥有“学习程式”，所以可以通过对话和教导让她‘成长’起来。",
}
`;

exports[`character > should get character casts 1`] = `
Object {
  "data": Array [
    Object {
      "actors": Array [],
      "subject": Object {
        "id": 12,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/c2/0a/12_aDoa8.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/c2/0a/12_aDoa8.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/c2/0a/12_aDoa8.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/c2/0a/12_aDoa8.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/c2/0a/12_aDoa8.jpg",
        },
        "info": "27话 / 2002年4月2日 / 浅香守生 / CLAMP / 阿部恒",
        "locked": false,
        "name": "ちょびっツ",
        "nameCN": "人形电脑天使心",
        "nsfw": false,
        "rating": Object {
          "count": Array [
            4,
            4,
            5,
            13,
            60,
            395,
            987,
            1226,
            369,
            168,
          ],
          "rank": 731,
          "score": 7.57,
          "total": 3231,
        },
        "type": 2,
      },
      "type": 1,
    },
  ],
  "total": 1,
}
`;

exports[`character > should get character collects 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": 1296496835,
      "user": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "group": 0,
        "id": 1,
        "joinedAt": 0,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "username": "1",
      },
    },
    Object {
      "createdAt": 1296496277,
      "user": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "group": 0,
        "id": 6162,
        "joinedAt": 0,
        "nickname": "nickname 6162",
        "sign": "sing 6162",
        "username": "6162",
      },
    },
  ],
  "total": 2,
}
`;
